polkadot/node/core/approval-voting-parallel/src/lib.rs

```diff
@@ -0,0 +1,950 @@
+					Err(err) => {
+						gum::info!(target: LOG_TARGET, ?err, "Approval voting parallel subsystem received an error");
+						return Err(err);
+					}
+				};
+				if !subsystem_enabled {
+					gum::trace!(target: LOG_TARGET, ?next_msg, "Parallel processing is not enabled skipping message");
+					continue;
+				}
+
+				match next_msg {
+					FromOrchestra::Signal(msg) => {
+						if matches!(msg, OverseerSignal::ActiveLeaves(_)) {
+							metrics_watcher.collect_metrics();
+						}
+
+						for worker in to_approval_distribution_workers.iter_mut() {
+							worker
+								.send_signal(msg.clone()).await?;
+						}
+
+						to_approval_voting_worker.send_signal(msg.clone()).await?;
+						if matches!(msg, OverseerSignal::Conclude) {
+							break;
+						}
+					},
+					FromOrchestra::Communication { msg } => match msg {
+						// The message the approval voting subsystem would've handled.
+						ApprovalVotingParallelMessage::ApprovedAncestor(_, _,_) |
+						ApprovalVotingParallelMessage::GetApprovalSignaturesForCandidate(_, _)  => {
+							// Safe to unwrap because we know the message is of the right type.
+							to_approval_voting_worker.send_message(msg.try_into().unwrap()).await;
+						},
+						// Now the message the approval distribution subsystem would've handled and need to
+						// be forwarded to the workers.
+						ApprovalVotingParallelMessage::NewBlocks(msg) => {
+							for worker in to_approval_distribution_workers.iter_mut() {
+								worker
+									.send_message(
+										ApprovalDistributionMessage::NewBlocks(msg.clone()),
+									)
+									.await;
+							}
+						},
+						ApprovalVotingParallelMessage::DistributeAssignment(assignment, claimed) => {
+							let worker_index = assignment.validator.0 as usize % to_approval_distribution_workers.len();
+							let worker = to_approval_distribution_workers.get_mut(worker_index).expect("Worker index is obtained modulo len; qed");
+							worker
+								.send_message(
+									ApprovalDistributionMessage::DistributeAssignment(assignment, claimed)
+								)
+								.await;
+
+						},
+						ApprovalVotingParallelMessage::DistributeApproval(vote) => {
+							let worker_index = vote.validator.0 as usize % to_approval_distribution_workers.len();
+							let worker = to_approval_distribution_workers.get_mut(worker_index).expect("Worker index is obtained modulo len; qed");
+							worker
+								.send_message(
+									ApprovalDistributionMessage::DistributeApproval(vote)
+								).await;
+
+						},
+						ApprovalVotingParallelMessage::NetworkBridgeUpdate(msg) => {
+							if let polkadot_node_subsystem::messages::NetworkBridgeEvent::PeerMessage(
+								peer_id,
+								msg,
+							) = msg
+							{
+								let (all_msgs_from_same_validator, messages_split_by_validator) = validator_index_for_msg(msg);
+
+								for (validator_index, msg) in all_msgs_from_same_validator.into_iter().chain(messages_split_by_validator.into_iter().flatten()) {
```

Comment:
Because the lifetime would be the same if you look at the returned messages, some of them might actually be allocated in the function while other are actually moved from the messages.